src/network/networkd-wwan.c

```diff
@@ -0,0 +1,666 @@
+        address->prefixlen = prefixlen;
+
+        if (address_get(link, address, &existing) < 0) /* The address is new. */
+                link->bearer_configured = false;
+        else
+                address_unmark(existing);
+
+        r = link_request_address(link, address, &link->bearer_messages,
+                                 bearer_address_handler, NULL);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to request address provided by bearer: %m");
+
+        return 0;
+}
+
+static int bearer_route_handler(
+                sd_netlink *rtnl,
+                sd_netlink_message *m,
+                Request *req,
+                Link *link,
+                Route *route) {
+
+        int r;
+
+        assert(link);
+
+        r = route_configure_handler_internal(m, req, route);
+        if (r <= 0)
+                return r;
+
+        if (link->bearer_messages == 0) {
+                link->bearer_configured = true;
+                link_check_ready(link);
+        }
+
+        return 0;
+}
+
+static int link_request_bearer_route(
+                Link *link,
+                int family,
+                const union in_addr_union *gw,
+                const union in_addr_union *prefsrc) {
+
+        _cleanup_(route_unrefp) Route *route = NULL;
+        Route *existing;
+        int r;
+
+        assert(link);
+        assert(IN_SET(family, AF_INET, AF_INET6));
+        assert(gw);
+        assert(prefsrc);
+
+        if (!in_addr_is_set(family, gw))
+                return 0;
+
+        /*
+         * FIXME: Allow not setting the default route via
+         * DHCPv4.UseGateway option
+         */
+        if (link->network->dhcp_use_gateway == 0)
+                return 0;
+
+        r = route_new(&route);
+        if (r < 0)
+                return log_oom();
+
+        route->source = NETWORK_CONFIG_SOURCE_MODEM_MANAGER;
+        route->family = family;
+        route->nexthop.family = family;
+        route->nexthop.gw = *gw;
+        if (link->network->priority != LINK_BRIDGE_PORT_PRIORITY_INVALID) {
+                route->priority = link->network->priority;
+                route->priority_set = true;
+        } else if (link->network->dhcp_route_metric != DHCP_ROUTE_METRIC) {
+                route->priority = link->network->dhcp_route_metric;
+                route->priority_set = true;
+        }
+
+        if (prefsrc)
+                route->prefsrc = *prefsrc;
+
+        if (route_get(link->manager, route, &existing) < 0) /* This is a new route. */
+
+                link->bearer_configured = false;
+        else
+                route_unmark(existing);
+
+        r = link_request_route(link, route, &link->bearer_messages, bearer_route_handler);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to request gateway provided by bearer: %m");
+
+        return 0;
+}
+
+static int link_apply_bearer_dns(Link *link, size_t n_dns, struct in_addr_data *dns) {
```

Comment:
Please do not clear `link->dns` and copy the dns addresses. This is set by DBus method by user request, and should have higher priority than acquired ones.

Instead, please update `networkd-state-file.c` and `networkd-json.c`.

Reply:
So, there are several ways to configure DNS (or several other, like NTP) servers:
- configure statically through DNS= setting in .network file,
- configure dynamically through DHCP or NDisc,
- configure transiently through DBus method.

The highest priority is DBus method. When specified through DBus, DNS servers specified by .network file and DHCP/NDisc are ignored.
When not specified by DBus, all other configurations are concatenated.

DNS servers obtained by MM should be handled as the same as DHCP or so.

See link_put_dns() (for manager state) and link_save() (for per-link state) in networkd-state-file.c.
Also, the same logic is used in dns_append_json() in networkd-json.c.

Reply:
Yeah. Exactly.